File: android/app/src/main/java/com/oppzippy/openscq30/lib/wrapper/SoundModes.kt
package com.oppzippy.openscq30.lib.wrapper

import com.oppzippy.openscq30.lib.protobuf.ambientSoundModeCycle
import com.oppzippy.openscq30.lib.protobuf.soundModes

data class SoundModes(
    val ambientSoundMode: AmbientSoundMode,
    val noiseCancelingMode: NoiseCancelingMode,
    val transparencyMode: TransparencyMode,
    val customNoiseCanceling: UByte,
) {
    fun toProtobuf(): com.oppzippy.openscq30.lib.protobuf.SoundModes {
        return soundModes {
            ambientSoundMode = this@SoundModes.ambientSoundMode.toProtobuf()
            noiseCancelingMode = this@SoundModes.noiseCancelingMode.toProtobuf()
            transparencyMode = this@SoundModes.transparencyMode.toProtobuf()
            customNoiseCanceling = this@SoundModes.customNoiseCanceling.toInt()
        }
    }
}

fun com.oppzippy.openscq30.lib.protobuf.SoundModes.toKotlin(): SoundModes {
    return SoundModes(
        ambientSoundMode = ambientSoundMode.toKotlin(),
        noiseCancelingMode = noiseCancelingMode.toKotlin(),
        transparencyMode = transparencyMode.toKotlin(),
        customNoiseCanceling = customNoiseCanceling.toUByte(),
    )
}

enum class AmbientSoundMode {
    Normal,
    Transparency,
    NoiseCanceling,
    ;

    fun toProtobuf(): com.oppzippy.openscq30.lib.protobuf.AmbientSoundMode {
        return when (this) {
            Normal -> com.oppzippy.openscq30.lib.protobuf.AmbientSoundMode.NORMAL
            Transparency -> com.oppzippy.openscq30.lib.protobuf.AmbientSoundMode.TRANSPARENCY
            NoiseCanceling -> com.oppzippy.openscq30.lib.protobuf.AmbientSoundMode.NOISE_CANCELING
        }
    }
}

fun com.oppzippy.openscq30.lib.protobuf.AmbientSoundMode.toKotlin(): AmbientSoundMode {
    return when (this) {
        com.oppzippy.openscq30.lib.protobuf.AmbientSoundMode.NORMAL -> AmbientSoundMode.Normal
        com.oppzippy.openscq30.lib.protobuf.AmbientSoundMode.TRANSPARENCY -> AmbientSoundMode.Transparency
        com.oppzippy.openscq30.lib.protobuf.AmbientSoundMode.NOISE_CANCELING -> AmbientSoundMode.NoiseCanceling
    }
}

data class AmbientSoundModeCycle(
    val normalMode: Boolean,
    val transparencyMode: Boolean,
    val noiseCancelingMode: Boolean,
) {
    fun toProtobuf(): com.oppzippy.openscq30.lib.protobuf.AmbientSoundModeCycle {
        return ambientSoundModeCycle {
            normalMode = this@AmbientSoundModeCycle.normalMode
            transparencyMode = this@AmbientSoundModeCycle.transparencyMode
            noiseCancelingMode = this@AmbientSoundModeCycle.noiseCancelingMode
        }
    }
}

fun com.oppzippy.openscq30.lib.protobuf.AmbientSoundModeCycle.toKotlin(): AmbientSoundModeCycle {
    return AmbientSoundModeCycle(
        normalMode = normalMode,
        transparencyMode = transparencyMode,
        noiseCancelingMode = noiseCancelingMode,
    )
}

enum class NoiseCancelingMode {
    Indoor,
    Outdoor,
    Transport,
    Custom,
    ;

    fun toProtobuf(): com.oppzippy.openscq30.lib.protobuf.NoiseCancelingMode {
        return when (this) {
            Indoor -> com.oppzippy.openscq30.lib.protobuf.NoiseCancelingMode.INDOOR
            Outdoor -> com.oppzippy.openscq30.lib.protobuf.NoiseCancelingMode.OUTDOOR
            Transport -> com.oppzippy.openscq30.lib.protobuf.NoiseCancelingMode.TRANSPORT
            Custom -> com.oppzippy.openscq30.lib.protobuf.NoiseCancelingMode.CUSTOM
        }
    }
}

fun com.oppzippy.openscq30.lib.protobuf.NoiseCancelingMode.toKotlin(): NoiseCancelingMode {
    return when (this) {
        com.oppzippy.openscq30.lib.protobuf.NoiseCancelingMode.INDOOR -> NoiseCancelingMode.Indoor
        com.oppzippy.openscq30.lib.protobuf.NoiseCancelingMode.OUTDOOR -> NoiseCancelingMode.Outdoor
        com.oppzippy.openscq30.lib.protobuf.NoiseCancelingMode.TRANSPORT -> NoiseCancelingMode.Transport
        com.oppzippy.openscq30.lib.protobuf.NoiseCancelingMode.CUSTOM -> NoiseCancelingMode.Custom
    }
}

enum class TransparencyMode {
    FullyTransparent,
    VocalMode,
    ;

    fun toProtobuf(): com.oppzippy.openscq30.lib.protobuf.TransparencyMode {
        return when (this) {
            FullyTransparent -> com.oppzippy.openscq30.lib.protobuf.TransparencyMode.FULLY_TRANSPARENT
            VocalMode -> com.oppzippy.openscq30.lib.protobuf.TransparencyMode.VOCAL_MODE
        }
    }
}

fun com.oppzippy.openscq30.lib.protobuf.TransparencyMode.toKotlin(): TransparencyMode {
    return when (this) {
        com.oppzippy.openscq30.lib.protobuf.TransparencyMode.FULLY_TRANSPARENT -> TransparencyMode.FullyTransparent
        com.oppzippy.openscq30.lib.protobuf.TransparencyMode.VOCAL_MODE -> TransparencyMode.VocalMode
    }
}

File: lib/src/devices/soundcore/common/structures/sound_modes.rs
use nom::{
    IResult, Parser,
    combinator::map,
    error::{ContextError, ParseError, context},
    number::complete::le_u8,
};

use crate::devices::soundcore::common::macros::sound_mode_enum;

#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, Default)]
pub struct SoundModes {
    pub ambient_sound_mode: AmbientSoundMode,
    pub noise_canceling_mode: NoiseCancelingMode,
    pub transparency_mode: TransparencyMode,
    pub custom_noise_canceling: CustomNoiseCanceling,
}

impl SoundModes {
    pub fn take<'a, E: ParseError<&'a [u8]> + ContextError<&'a [u8]>>(
        input: &'a [u8],
    ) -> IResult<&'a [u8], Self, E> {
        context(
            "group of sound modes",
            map(
                (
                    AmbientSoundMode::take,
                    NoiseCancelingMode::take,
                    TransparencyMode::take,
                    CustomNoiseCanceling::take,
                ),
                |(
                    ambient_sound_mode,
                    noise_canceling_mode,
                    transparency_mode,
                    custom_noise_canceling,
                )| {
                    Self {
                        ambient_sound_mode,
                        noise_canceling_mode,
                        transparency_mode,
                        custom_noise_canceling,
                    }
                },
            ),
        )
        .parse_complete(input)
    }

    pub fn bytes(&self) -> [u8; 4] {
        [
            self.ambient_sound_mode.byte(),
            self.noise_canceling_mode.byte(),
            self.transparency_mode.byte(),
            self.custom_noise_canceling.value(),
        ]
    }
}

sound_mode_enum!(
    pub enum AmbientSoundMode {
        NoiseCanceling = 0,
        Transparency = 1,
        Normal = 2,
    }
);

sound_mode_enum!(
    pub enum NoiseCancelingMode {
        Transport = 0,
        Outdoor = 1,
        Indoor = 2,
        Custom = 3,
    }
);

sound_mode_enum!(
    pub enum TransparencyMode {
        FullyTransparent = 0,
        VocalMode = 1,
    }
);

#[derive(Debug, Clone, Copy, Default, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct CustomNoiseCanceling {
    value: u8,
}

impl CustomNoiseCanceling {
    pub fn new(value: u8) -> Self {
        // Not sure what 255 means here, but it is allowed in addition to 0-10
        let clamped_value = if value == 255 {
            value
        } else {
            value.clamp(0, 10)
        };
        Self {
            value: clamped_value,
        }
    }

    pub fn value(&self) -> u8 {
        self.value
    }

    pub fn take<'a, E: ParseError<&'a [u8]> + ContextError<&'a [u8]>>(
        input: &'a [u8],
    ) -> IResult<&'a [u8], Self, E> {
        context(
            "custom noise canceling",
            map(le_u8, |custom_noise_canceling_level| {
                Self::new(custom_noise_canceling_level)
            }),
        )
        .parse_complete(input)
    }
}
